File: OOYA_SP/C#/OPLab/ITCompany.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace OPLab
{
    public class ITCompany
    {
        private string _name;
        private List<Programmer> _programmers;

        public ITCompany()
        { }

        public ITCompany(string name, List<Programmer> programmers)
        {
            _name = name;
            _programmers = programmers;
        }

        ~ITCompany()
        {
            //destructor
            Console.WriteLine("Company was deleted.");
        }

        public string Name
        {
            get => _name;
            set => _name = value;
        }

        public List<Programmer> Programmers
        {
            get => _programmers;
            set => _programmers = value;
        }

        public void Sort()
        {
            _programmers =  _programmers.OrderBy(p => p.Position ).ThenBy(p => p.Name).ThenBy(p => p.Surname).ToList();
        }

        public void Print()
        {
            Console.WriteLine("IT Company: {0}", _name);
            Console.WriteLine();
            foreach (Programmer programmer in _programmers)
            {
                programmer.Print();
            }
        }

        Position getPosition(string s)
        {
            if (s == "JUNIOR")
            {
                return Position.JUNIOR;
            }
            else if (s == "MIDDLE")
            {
                return Position.MIDDLE;
            }
            else
            {
                return Position.SENIOR;
            }
        }

        List<string> getWords(string s)
        {
            List<string> result = new List<string>();
            int start = 0;
            int end = s.IndexOf(" ");
            while (end > 0)
            {
                var a = s.Substring(start, end - start);
                result.Add(a);
                start = end + 1;
                end = s.IndexOf(" ", start);
            }
            //result.Add(s.Substring(start, end));
            return result;
        }

        public ITCompany ReadInfoFromFile()
        {
            string path = "input.txt";
           
                using (StreamReader sr = File.OpenText(path))
                {
                    string nameCompany = sr.ReadLine();
                    string tmp = sr.ReadLine();
                    int programmersQuant = Convert.ToInt32(tmp);
                    sr.ReadLine();
                    var allProgrammers = new List<Programmer>();
                    for (int i = 0; i < programmersQuant; i++)
                    {
                        string type = sr.ReadLine();
                        string name = sr.ReadLine();
                        string surname = sr.ReadLine();
                        Position position = getPosition(sr.ReadLine());
                        List<string> skills = getWords(sr.ReadLine());
                        if (type == "Developer")
                        {
                            string knowledgeArea = sr.ReadLine();
                            string currentProject = sr.ReadLine();
                            allProgrammers.Add(new Developer(name, surname, position, skills, knowledgeArea, currentProject));
                        }
                        else if (type == "Manager")
                        {
                            List<string> projects = getWords(sr.ReadLine());
                            int subordinatesQuantity = Convert.ToInt32(sr.ReadLine());
                            allProgrammers.Add(new Manager(name, surname, position, skills, subordinatesQuantity, projects));
                        }
                    // sr.ReadLine();
                    }
                    var result = new ITCompany(nameCompany, allProgrammers);
                    return result;
                }
        }

        public void DeleteNotOdd()
        {
            List<Programmer> programmers = new List<Programmer>();
            for (int i = 0; i < this._programmers.Count(); i++)
            {
                if (this._programmers[i].GetType() == typeof(Developer))
                {
                    var developer = (Developer)Convert.ChangeType(this.Programmers[i], typeof(Developer));
                    if (Convert.ToInt32(developer.CurrentProject[developer.CurrentProject.Length - 2])%2 != 0)
                    {
                        programmers.Add(developer);
                    }
                }
                else
                {
                    programmers.Add(this._programmers[i]);
                }
            }
			this._programmers = null;
			this._programmers = programmers;
        }
    }

}

File: OOYA_SP/Python/ITCompany.py
from Position import Position
from Developer import Developer
from Manager import Manager

class ITCompany:
    def __new__(cls):
        return super(ITCompany, cls).__new__(cls)

    def __init__(self, name, programmers):
        self._name = name
        self._programmers = programmers

    def __del__(self):
        print("Company deleted")

    def __print__(self):
        print(self._name)
        for person in self._programmers :
            person.__print__();

    def getPosition(self, str):
        if str == "JUNIOR" :
            return Position.JUNIOR
        elif  str == "MIDDLE" :
                return Position.MIDDLE
        else : return Position.SENIOR;


    def __read__(self):
        f = open('input.txt', 'r')
        name = f.readline().replace('\n','')
        quant = int(f.readline())
        f.readline()
        programmers = []

        for i  in range(quant) :
            type = str(f.readline()).replace('\n','')
            nm = f.readline().replace('\n','')
            srnm = f.readline().replace('\n','')
            pos = self.getPosition(f.readline())
            skls = f.readline().split()
            if type == "Developer" :
                knwldg = f.readline().replace('\n','')
                prjcts = f.readline().split()
                dev = Developer(nm, srnm, pos, skls, knwldg, prjcts)
                programmers.append(dev)
            else:
                prjcts = f.readline().split()
                quant = int(f.readline())
                manager = Manager(nm, srnm,pos, skls, quant, prjcts)
                programmers.append(manager)

        cm = ITCompany.__new__(ITCompany)
        cm.__init__(name, programmers)
        return cm

    def __sort__(self):
        sorted(self._programmers, key=lambda p : (p._position, p._name, p._surname))

    def __removeNotOdd__(self):
        programmers = []
        for programmer in self._programmers :
             if type(programmer) is Developer :
                 if int(programmer._currentProject[0][len(programmer._currentProject[0]) - 1]) % 2 !=0 :
                     programmers.append(programmer)
             else :
                 programmers.append(programmer)
        self._programmers = programmers








